# ETL Pipeline Preparation
Follow the instructions below to help you create your ETL pipeline.
### 1. Import libraries and load datasets.
- Import Python libraries
- Load `messages.csv` into a dataframe and inspect the first few lines.
- Load `categories.csv` into a dataframe and inspect the first few lines.

In [2]:
# import libraries
import sqlite3
from sqlite3 import Error
import pandas as pd
from sqlalchemy import create_engine

In [44]:
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', -1)

In [4]:
# load messages dataset
messages = pd.read_csv('messages.csv', dtype=str)
messages.head()

id                                            message                                           original   genre
0   2  Weather update - a cold front from Cuba that c...  Un front froid se retrouve sur Cuba ce matin. ...  direct
1   7            Is the Hurricane over or is it not over                 Cyclone nan fini osinon li pa fini  direct
2   8                    Looking for someone but no name  Patnm, di Maryani relem pou li banm nouvel li ...  direct
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...  UN reports Leogane 80-90 destroyed. Only Hospi...  direct
4  12  says: west side of Haiti, rest of the country ...  facade ouest d Haiti et le reste du pays aujou...  direct

In [5]:
messages.shape

(26248, 4)

In [6]:
# load categories dataset
categories = pd.read_csv('categories.csv', dtype=str)
categories.head()

id                                         categories
0   2  related-1;request-0;offer-0;aid_related-0;medi...
1   7  related-1;request-0;offer-0;aid_related-1;medi...
2   8  related-1;request-0;offer-0;aid_related-0;medi...
3   9  related-1;request-1;offer-0;aid_related-1;medi...
4  12  related-1;request-0;offer-0;aid_related-0;medi...

In [7]:
categories.shape

(26248, 2)

In [8]:
messages.isnull().sum()

id              0
message         0
original    16064
genre           0
dtype: int64

In [9]:
categories.isnull().sum()

id            0
categories    0
dtype: int64

### 2. Merge datasets.
- Merge the messages and categories datasets using the common id
- Assign this combined dataset to `df`, which will be cleaned in the following steps

In [10]:
# merge datasets
# TODO: merge the data frames together based on their common columns
# in this case, the common columns is id
df = pd.merge(messages,categories, on=['id'])
df.head()

id                                            message                                           original   genre                                         categories
0   2  Weather update - a cold front from Cuba that c...  Un front froid se retrouve sur Cuba ce matin. ...  direct  related-1;request-0;offer-0;aid_related-0;medi...
1   7            Is the Hurricane over or is it not over                 Cyclone nan fini osinon li pa fini  direct  related-1;request-0;offer-0;aid_related-1;medi...
2   8                    Looking for someone but no name  Patnm, di Maryani relem pou li banm nouvel li ...  direct  related-1;request-0;offer-0;aid_related-0;medi...
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...  UN reports Leogane 80-90 destroyed. Only Hospi...  direct  related-1;request-1;offer-0;aid_related-1;medi...
4  12  says: west side of Haiti, rest of the country ...  facade ouest d Haiti et le reste du pays aujou...  direct  related-1;request-0;offer-0;aid_related-0;medi...

In [11]:
df.shape

(26386, 5)

- 26248 in each inputs
- 26386 in final output
- Looks like there are duplicates on the field id

In [12]:
# Select all duplicate rows based on key column 'id'
duplicatecategories = categories[categories.duplicated(['id'])]
duplicatecategories.shape

(68, 2)

In [13]:
duplicatecategories.head()

id                                         categories
163    202  related-1;request-1;offer-0;aid_related-1;medi...
656    804  related-2;request-0;offer-0;aid_related-0;medi...
710    862  related-1;request-0;offer-0;aid_related-0;medi...
1408  1652  related-1;request-1;offer-0;aid_related-1;medi...
2131  2446  related-1;request-0;offer-0;aid_related-0;medi...

In [14]:
categories[categories['id']=="202"]

id                                         categories
162  202  related-1;request-1;offer-0;aid_related-1;medi...
163  202  related-1;request-1;offer-0;aid_related-1;medi...

In [15]:
duplicatemessages = messages[messages.duplicated(['id'])]
duplicatemessages.shape

(68, 4)

In [16]:
messages[messages['id']=="202"]

id                                            message                                           original   genre
162  202  ?? port au prince ?? and food. they need gover...  p bay pap la syen ak manje. Yo bezwen ed gouve...  direct
163  202  ?? port au prince ?? and food. they need gover...  p bay pap la syen ak manje. Yo bezwen ed gouve...  direct

### 3. Split `categories` into separate category columns.
- Split the values in the `categories` column on the `;` character so that each value becomes a separate column. You'll find [this method](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.Series.str.split.html) very helpful! Make sure to set `expand=True`.
- Use the first row of categories dataframe to create column names for the categories data.
- Rename columns of `categories` with new column names.

In [17]:
categories = categories['categories'].str.split(';',expand=True)
categories.head()

0          1        2              3               4                   5                    6           7           8              9        10      11         12          13       14                15          16       17           18                        19           20           21             22       23           24       25             26                      27                 28        29       30      31            32      33               34               35
0  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
1  related-1  request-0  offer-0  aid_related-1  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-1  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
2  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
3  related-1  request-1  offer-0  aid_related-1  medical_help-0  medical_products-1  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-1  infrastructure_related-1  transport-0  buildings-1  electricity-0  tools-0  hospitals-1  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
4  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0

In [18]:
categories.tail()

0          1        2              3               4                   5                    6           7           8              9        10      11         12          13       14                15          16       17           18                        19           20           21             22       23           24       25             26                      27                 28        29       30      31            32      33               34               35
26243  related-0  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
26244  related-0  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
26245  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
26246  related-1  request-0  offer-0  aid_related-1  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-1  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
26247  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0

In [19]:
row = categories.head(1)
row

0          1        2              3               4                   5                    6           7           8              9        10      11         12          13       14                15          16       17           18                        19           20           21             22       23           24       25             26                      27                 28        29       30      31            32      33               34               35
0  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0

In [20]:
categories.columns = categories.iloc[0].apply(lambda x : x[:-2])


In [21]:
categories.head()

0    related    request    offer    aid_related    medical_help    medical_products    search_and_rescue    security    military    child_alone    water    food    shelter    clothing    money    missing_people    refugees    death    other_aid    infrastructure_related    transport    buildings    electricity    tools    hospitals    shops    aid_centers    other_infrastructure    weather_related    floods    storm    fire    earthquake    cold    other_weather    direct_report
0  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
1  related-1  request-0  offer-0  aid_related-1  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-1  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
2  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
3  related-1  request-1  offer-0  aid_related-1  medical_help-0  medical_products-1  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-1  infrastructure_related-1  transport-0  buildings-1  electricity-0  tools-0  hospitals-1  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
4  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0

### 4. Convert category values to just numbers 0 or 1.
- Iterate through the category columns in df to keep only the last character of each string (the 1 or 0). For example, `related-0` becomes `0`, `related-1` becomes `1`. Convert the string to a numeric value.
- You can perform [normal string actions on Pandas Series](https://pandas.pydata.org/pandas-docs/stable/text.html#indexing-with-str), like indexing, by including `.str` after the Series. You may need to first convert the Series to be of type string, which you can do with `astype(str)`.

In [22]:
for column in categories:
    # set each value to be the last character of the string
    categories[column] = categories[column].apply(lambda x : x[-1:])
    
    # convert column from string to numeric
    categories[column] =  categories[column].astype(int)
categories.head()

0  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
0        1        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0              0
1        1        0      0            1             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          1                       0          0          0            0      0          0      0            0                     0                1       0      1     0           0     0              0              0
2        1        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0              0
3        1        1      0            1             0                 1                  0         0         0            0      0     0        0         0      0               0         0      0          1                       1          0          1            0      0          1      0            0                     0                0       0      0     0           0     0              0              0
4        1        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0              0

In [23]:
categories.replace(2, 1, inplace=True)

### 5. Replace `categories` column in `df` with new category columns.
- Drop the categories column from the df dataframe since it is no longer needed.
- Concatenate df and categories data frames.

In [24]:
# drop the original categories column from `df`

df.drop(['categories'], axis=1, inplace=True)
df.head()

id                                            message                                           original   genre
0   2  Weather update - a cold front from Cuba that c...  Un front froid se retrouve sur Cuba ce matin. ...  direct
1   7            Is the Hurricane over or is it not over                 Cyclone nan fini osinon li pa fini  direct
2   8                    Looking for someone but no name  Patnm, di Maryani relem pou li banm nouvel li ...  direct
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...  UN reports Leogane 80-90 destroyed. Only Hospi...  direct
4  12  says: west side of Haiti, rest of the country ...  facade ouest d Haiti et le reste du pays aujou...  direct

In [25]:
# concatenate the original dataframe with the new `categories` dataframe
df = pd.concat([df, categories], axis=1)
df.head()

id                                            message                                           original   genre  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
0   2  Weather update - a cold front from Cuba that c...  Un front froid se retrouve sur Cuba ce matin. ...  direct      1.0      0.0    0.0          0.0           0.0               0.0                0.0       0.0       0.0          0.0    0.0   0.0      0.0       0.0    0.0             0.0       0.0    0.0        0.0                     0.0        0.0        0.0          0.0    0.0        0.0    0.0          0.0                   0.0              0.0     0.0    0.0   0.0         0.0   0.0            0.0            0.0
1   7            Is the Hurricane over or is it not over                 Cyclone nan fini osinon li pa fini  direct      1.0      0.0    0.0          1.0           0.0               0.0                0.0       0.0       0.0          0.0    0.0   0.0      0.0       0.0    0.0             0.0       0.0    0.0        1.0                     0.0        0.0        0.0          0.0    0.0        0.0    0.0          0.0                   0.0              1.0     0.0    1.0   0.0         0.0   0.0            0.0            0.0
2   8                    Looking for someone but no name  Patnm, di Maryani relem pou li banm nouvel li ...  direct      1.0      0.0    0.0          0.0           0.0               0.0                0.0       0.0       0.0          0.0    0.0   0.0      0.0       0.0    0.0             0.0       0.0    0.0        0.0                     0.0        0.0        0.0          0.0    0.0        0.0    0.0          0.0                   0.0              0.0     0.0    0.0   0.0         0.0   0.0            0.0            0.0
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...  UN reports Leogane 80-90 destroyed. Only Hospi...  direct      1.0      1.0    0.0          1.0           0.0               1.0                0.0       0.0       0.0          0.0    0.0   0.0      0.0       0.0    0.0             0.0       0.0    0.0        1.0                     1.0        0.0        1.0          0.0    0.0        1.0    0.0          0.0                   0.0              0.0     0.0    0.0   0.0         0.0   0.0            0.0            0.0
4  12  says: west side of Haiti, rest of the country ...  facade ouest d Haiti et le reste du pays aujou...  direct      1.0      0.0    0.0          0.0           0.0               0.0                0.0       0.0       0.0          0.0    0.0   0.0      0.0       0.0    0.0             0.0       0.0    0.0        0.0                     0.0        0.0        0.0          0.0    0.0        0.0    0.0          0.0                   0.0              0.0     0.0    0.0   0.0         0.0   0.0            0.0            0.0

In [26]:
df.tail()

id                                            message original genre  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
26381  30261  The training demonstrated how to enhance micro...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN
26382  30262  A suitable candidate has been selected and OCH...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN
26383  30263  Proshika, operating in Cox's Bazar municipalit...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN
26384  30264  Some 2,000 women protesting against the conduc...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN
26385  30265  A radical shift in thinking came about as a re...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN

In [27]:
df.shape

(26386, 40)

### 6. Remove duplicates.
- Check how many duplicates are in this dataset.
- Drop the duplicates.
- Confirm duplicates were removed.

In [28]:
# check number of duplicates
df[df.duplicated(subset='message')].count()

id                        209
message                   209
original                   93
genre                     209
related                   209
request                   209
offer                     209
aid_related               209
medical_help              209
medical_products          209
search_and_rescue         209
security                  209
military                  209
child_alone               209
water                     209
food                      209
shelter                   209
clothing                  209
money                     209
missing_people            209
refugees                  209
death                     209
other_aid                 209
infrastructure_related    209
transport                 209
buildings                 209
electricity               209
tools                     209
hospitals                 209
shops                     209
aid_centers               209
other_infrastructure      209
weather_related           209
floods    

In [29]:
# drop duplicates
df.drop_duplicates(subset='message', inplace=True)

In [30]:
# check number of duplicates
df[df.duplicated(subset='message')].count()

id                        0
message                   0
original                  0
genre                     0
related                   0
request                   0
offer                     0
aid_related               0
medical_help              0
medical_products          0
search_and_rescue         0
security                  0
military                  0
child_alone               0
water                     0
food                      0
shelter                   0
clothing                  0
money                     0
missing_people            0
refugees                  0
death                     0
other_aid                 0
infrastructure_related    0
transport                 0
buildings                 0
electricity               0
tools                     0
hospitals                 0
shops                     0
aid_centers               0
other_infrastructure      0
weather_related           0
floods                    0
storm                     0
fire                

In [31]:
df.shape

(26177, 40)

In [32]:
df.head()

id                                            message                                           original   genre  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
0   2  Weather update - a cold front from Cuba that c...  Un front froid se retrouve sur Cuba ce matin. ...  direct      1.0      0.0    0.0          0.0           0.0               0.0                0.0       0.0       0.0          0.0    0.0   0.0      0.0       0.0    0.0             0.0       0.0    0.0        0.0                     0.0        0.0        0.0          0.0    0.0        0.0    0.0          0.0                   0.0              0.0     0.0    0.0   0.0         0.0   0.0            0.0            0.0
1   7            Is the Hurricane over or is it not over                 Cyclone nan fini osinon li pa fini  direct      1.0      0.0    0.0          1.0           0.0               0.0                0.0       0.0       0.0          0.0    0.0   0.0      0.0       0.0    0.0             0.0       0.0    0.0        1.0                     0.0        0.0        0.0          0.0    0.0        0.0    0.0          0.0                   0.0              1.0     0.0    1.0   0.0         0.0   0.0            0.0            0.0
2   8                    Looking for someone but no name  Patnm, di Maryani relem pou li banm nouvel li ...  direct      1.0      0.0    0.0          0.0           0.0               0.0                0.0       0.0       0.0          0.0    0.0   0.0      0.0       0.0    0.0             0.0       0.0    0.0        0.0                     0.0        0.0        0.0          0.0    0.0        0.0    0.0          0.0                   0.0              0.0     0.0    0.0   0.0         0.0   0.0            0.0            0.0
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...  UN reports Leogane 80-90 destroyed. Only Hospi...  direct      1.0      1.0    0.0          1.0           0.0               1.0                0.0       0.0       0.0          0.0    0.0   0.0      0.0       0.0    0.0             0.0       0.0    0.0        1.0                     1.0        0.0        1.0          0.0    0.0        1.0    0.0          0.0                   0.0              0.0     0.0    0.0   0.0         0.0   0.0            0.0            0.0
4  12  says: west side of Haiti, rest of the country ...  facade ouest d Haiti et le reste du pays aujou...  direct      1.0      0.0    0.0          0.0           0.0               0.0                0.0       0.0       0.0          0.0    0.0   0.0      0.0       0.0    0.0             0.0       0.0    0.0        0.0                     0.0        0.0        0.0          0.0    0.0        0.0    0.0          0.0                   0.0              0.0     0.0    0.0   0.0         0.0   0.0            0.0            0.0

In [33]:
df.tail()

id                                            message original genre  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
26381  30261  The training demonstrated how to enhance micro...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN
26382  30262  A suitable candidate has been selected and OCH...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN
26383  30263  Proshika, operating in Cox's Bazar municipalit...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN
26384  30264  Some 2,000 women protesting against the conduc...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN
26385  30265  A radical shift in thinking came about as a re...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN

In [34]:
messages[messages['id']=='30261']

id                                            message original genre
26243  30261  The training demonstrated how to enhance micro...      NaN  news

In [35]:
categories1 = pd.read_csv('categories.csv', dtype=str)
categories1[categories1['id']=='30261']

id                                         categories
26243  30261  related-0;request-0;offer-0;aid_related-0;medi...

In [36]:
df[df['id']=='30261']

id                                            message original genre  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
26381  30261  The training demonstrated how to enhance micro...      NaN  news      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN            NaN

In [46]:
df[df['id']=='30261']['message']

26381    The training demonstrated how to enhance micronutrient, protein and energy intake for meals using locally grown horticulture produce, rice and fish.
Name: message, dtype: object

In [45]:
df[df['id']=='30261']['message'].map(str).apply(len)

26381    148
Name: message, dtype: int64

In [47]:
df[df['message']=='The training demonstrated how to enhance micronutrient, protein and energy intake for meals using locally grown horticulture produce, rice and fish.']

id                                                                                                                                               message original genre  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
26381  30261  The training demonstrated how to enhance micronutrient, protein and energy intake for meals using locally grown horticulture produce, rice and fish.  NaN      news NaN      NaN      NaN    NaN          NaN           NaN               NaN                NaN       NaN       NaN          NaN    NaN   NaN      NaN       NaN    NaN             NaN       NaN    NaN        NaN                     NaN        NaN        NaN          NaN    NaN        NaN    NaN          NaN                   NaN              NaN     NaN    NaN   NaN         NaN   NaN            NaN

## ERROR ABOVE

### 7. Save the clean dataset into an sqlite database.
You can do this with pandas [`to_sql` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_sql.html) combined with the SQLAlchemy library. Remember to import SQLAlchemy's `create_engine` in the first cell of this notebook to use it below.

In [38]:
engine = create_engine('sqlite:///DisasterResponse.db')
df.to_sql('InsertTableName1', engine, index=False)

### 8. Use this notebook to complete `etl_pipeline.py`
Use the template file attached in the Resources folder to write a script that runs the steps above to create a database based on new datasets specified by the user. Alternatively, you can complete `etl_pipeline.py` in the classroom on the `Project Workspace IDE` coming later.

### My new Solution

In [37]:
# import libraries
import sqlite3
from sqlite3 import Error
import pandas as pd
from sqlalchemy import create_engine

In [38]:
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', -1)

In [39]:
newmessages = pd.read_csv('messages.csv', dtype=str)
newcategories = pd.read_csv('categories.csv', dtype=str)


In [40]:
newmessages.shape

(26248, 4)

In [41]:
newcategories.shape

(26248, 2)

In [42]:
i = newcategories.columns.get_loc('categories')
mycategories2 = newcategories['categories'].str.split(';',expand=True)
newcategories = pd.concat([newcategories.iloc[:, :i], mycategories2, newcategories.iloc[:, i+1:]], axis=1)

In [43]:
newcategories.shape

(26248, 37)

In [44]:
newcategories.head()

id          0          1        2              3               4                   5                    6           7           8              9       10      11         12          13       14                15          16       17           18                        19           20           21             22       23           24       25             26                      27                 28        29       30      31            32      33               34               35
0  2   related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
1  7   related-1  request-0  offer-0  aid_related-1  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-1  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
2  8   related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
3  9   related-1  request-1  offer-0  aid_related-1  medical_help-0  medical_products-1  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-1  infrastructure_related-1  transport-0  buildings-1  electricity-0  tools-0  hospitals-1  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
4  12  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0

In [52]:
#newcategories.columns = newcategories.iloc[0].apply(lambda x : x[:-2])
col_list = ['id']    
# newcategories.rename(columns=lambda x: x[:-2] if x not in col_list else x)
# newcategories.columns = newcategories.iloc[0].apply(lambda x : x[:-2] if x not in col_list else )
col_map = newcategories.iloc[0].apply(lambda x : x[:-2])
type(col_map)
col_map
col_map = col_map.replace('', 'id')
newcategories.columns = col_map

In [53]:
newcategories.head(2)

0 id    related    request    offer    aid_related    medical_help    medical_products    search_and_rescue    security    military    child_alone    water    food    shelter    clothing    money    missing_people    refugees    death    other_aid    infrastructure_related    transport    buildings    electricity    tools    hospitals    shops    aid_centers    other_infrastructure    weather_related    floods    storm    fire    earthquake    cold    other_weather    direct_report
0  2  related-1  request-0  offer-0  aid_related-0  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0
1  7  related-1  request-0  offer-0  aid_related-1  medical_help-0  medical_products-0  search_and_rescue-0  security-0  military-0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0  missing_people-0  refugees-0  death-0  other_aid-1  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0  weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0  other_weather-0  direct_report-0

In [54]:
for column in newcategories:
    # set each value to be the last character of the string
    if(column != 'id'):
        newcategories[column] = newcategories[column].apply(lambda x : x[-1:])
    
        # convert column from string to numeric
        newcategories[column] =  newcategories[column].astype(int)
newcategories.head()

0  id  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
0  2   1        0        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0            
1  7   1        0        0      1            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      1          0                       0          0          0            0      0          0      0            0                     1                0       1      0     0           0     0              0            
2  8   1        0        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0            
3  9   1        1        0      1            0             1                 0                  0         0         0            0      0     0        0         0      0               0         0      1          1                       0          1          0            0      1          0      0            0                     0                0       0      0     0           0     0              0            
4  12  1        0        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0

In [55]:
newmessages.head(3)

id                                                             message                                                                                                                                                   original   genre
0  2  Weather update - a cold front from Cuba that could pass over Haiti  Un front froid se retrouve sur Cuba ce matin. Il pourrait traverser Haiti demain. Des averses de pluie isolee sont encore prevues sur notre region ce soi  direct
1  7  Is the Hurricane over or is it not over                             Cyclone nan fini osinon li pa fini                                                                                                                         direct
2  8  Looking for someone but no name                                     Patnm, di Maryani relem pou li banm nouvel li ak timoun yo. Mesi se john jean depi Monben kwochi.                                                          direct

In [56]:
newmessages.dtypes

id          object
message     object
original    object
genre       object
dtype: object

In [57]:

#newmessages.astype({'id': 'int64'}).dtypes

In [58]:
newmessages.dtypes

id          object
message     object
original    object
genre       object
dtype: object

In [59]:
#newmessages['id'] = newmessages['id'].astype(int)

In [60]:
newmessages.dtypes

id          object
message     object
original    object
genre       object
dtype: object

In [61]:
newdf = pd.merge(newmessages,newcategories, on=['id'])

In [62]:
newdf.shape

(26386, 40)

In [63]:
newmessages.shape

(26248, 4)

In [64]:
newcategories.shape

(26248, 37)

In [65]:
newdf[newdf['id']=='30261']

id                                                                                                                                               message original genre  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
26381  30261  The training demonstrated how to enhance micronutrient, protein and energy intake for meals using locally grown horticulture produce, rice and fish.  NaN      news  0        0        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0

In [67]:
newmessages[newmessages['id']=='30261']

id                                                                                                                                               message original genre
26243  30261  The training demonstrated how to enhance micronutrient, protein and energy intake for meals using locally grown horticulture produce, rice and fish.  NaN      news

In [68]:
newcategories[newcategories['id']=='30261']

0         id  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
26243  30261  0        0        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0

In [69]:
newcategories.shape

(26248, 37)

In [70]:
newcategories.tail()

0         id  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
26243  30261  0        0        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0            
26244  30262  0        0        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0            
26245  30263  1        0        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0            
26246  30264  1        0        0      1            0             0                 0                  0         1         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0            
26247  30265  1        0        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0

In [71]:
newcategories['id'].unique()

array(['2', '7', '8', ..., '30263', '30264', '30265'], dtype=object)

In [72]:
newdf[newdf['id']=='30261']

id                                                                                                                                               message original genre  related  request  offer  aid_related  medical_help  medical_products  search_and_rescue  security  military  child_alone  water  food  shelter  clothing  money  missing_people  refugees  death  other_aid  infrastructure_related  transport  buildings  electricity  tools  hospitals  shops  aid_centers  other_infrastructure  weather_related  floods  storm  fire  earthquake  cold  other_weather  direct_report
26381  30261  The training demonstrated how to enhance micronutrient, protein and energy intake for meals using locally grown horticulture produce, rice and fish.  NaN      news  0        0        0      0            0             0                 0                  0         0         0            0      0     0        0         0      0               0         0      0          0                       0          0          0            0      0          0      0            0                     0                0       0      0     0           0     0              0

In [73]:
newdf.shape

(26386, 40)

In [74]:
# Check number of duplicates before removal of duplicates
print('Number of duplicates before removal are: {}'.format(sum(newdf.duplicated())))

Number of duplicates before removal are: 170


In [75]:
# Drop duplicates
newdf.drop_duplicates(inplace=True)

In [76]:
# Check number of duplicates after removal of duplicates
print('Number of duplicates after removal are: {}'.format(sum(newdf.duplicated())))

Number of duplicates after removal are: 0


In [77]:
newdf.shape

(26216, 40)

In [78]:
engine = create_engine('sqlite:///DisasterResponse.db')
newdf.to_sql('Labelled_table', engine, index=False, if_exists='replace')